contracts/shadeswap_market_oracle/src/contract.rs

```diff
@@ -0,0 +1,247 @@
+    deps: &mut Extern<S, A, Q>,
+    env: &Env,
+    admins: Option<Vec<HumanAddr>>,
+    router: Option<Contract>,
+) -> StdResult<HandleResponse> {
+
+    let mut config = CONFIG.load(&deps.storage)?;
+
+    if !config.admins.contains(&env.message.sender) {
+        return Err(StdError::unauthorized());
+    }
+
+    if let Some(admins) = admins {
+        if !admins.is_empty() {
+            config.admins = admins;
+        }
+    }
+
+    if let Some(router) = router {
+        config.router = router;
+    }
+
+    CONFIG.save(&mut deps.storage, &config)?;
+
+    Ok(HandleResponse {
+        messages: vec![],
+        log: vec![],
+        data: Some(to_binary(&HandleAnswer::UpdateConfig { 
+            status: ResponseStatus::Success, 
+        })?),
+    })
+}
+
+pub fn query<S: Storage, A: Api, Q: Querier>(deps: &Extern<S, A, Q>, msg: QueryMsg) -> QueryResult {
+    pad_query_result(
+        match msg {
+            QueryMsg::GetConfig {} => to_binary(&CONFIG.load(&deps.storage)?),
+            QueryMsg::GetPrice { key } => to_binary(&try_query_price(deps, key)?),
+            QueryMsg::GetPrices { keys } => {
+
+                let mut prices = vec![];
+                for key in keys {
+                    prices.push(try_query_price(deps, key)?);
+                }
+                to_binary(&prices)
+            },
+        },
+        BLOCK_SIZE,
+    )
+}
+
+fn try_query_price<S: Storage, A: Api, Q: Querier>(
+    deps: &Extern<S, A, Q>,
+    key: String,
+) -> StdResult<OraclePrice> {
+    let config = CONFIG.load(&deps.storage)?;
+
+    let primary_token: Contract = PRIMARY_TOKEN.load(&deps.storage)?;
+    //return Err(StdError::generic_err("here"));
+    //let base_token = BASE_TOKEN.load(&deps.storage)?;
+
+    let primary_info = PRIMARY_INFO.load(&deps.storage)?;
+
+    // Simulate trade 1 primary -> 1 base
+    let sim: EstimatedPriceResponse = ShadeSwapQueryMsg::GetEstimatedPrice {
+        offer: TokenAmount {
+            amount: Uint128(10u128.pow(primary_info.decimals.into())),
+            token: TokenType::CustomToken {
+                contract_addr: primary_token.address,
+                token_code_hash: primary_token.code_hash,
+            },
+        }
+    }.query(
+        &deps.querier,
+        config.pair.code_hash.clone(),
+        config.pair.address.clone(),
+    )?;
+
+    // Normalize to 'rate * 10^18'
+    let base_info = BASE_INFO.load(&deps.storage)?;
+    let exchange_rate = normalize_price(sim.estimated_price, base_info.decimals);
+
+    // Query router for base_peg/USD
+    let oracle = query_oracle(&config.router, &deps.querier, config.base_peg.clone())?;
+    let base_usd_price = query_price(&oracle, &deps.querier, config.base_peg.clone())?;
+
+    // Translate price to primary/USD
+    let price = base_usd_price.price.rate.multiply_ratio(exchange_rate, 10u128.pow(18));
+
+    Ok(OraclePrice::new(key,
+        ReferenceData {
+            rate: price,
+            last_updated_base: 0,
```

Comment:
shouldn't these times be based on what time is returned from the base_usd_price query on line 235? otherwise the consumer has no information about whether the queried price is stale or not

Reply:
Yeah I need to add that in, the sienna market oracle probably missed it as well